### ISSUE TITLE
Webhook authentication

### ISSUE THREAD

**Issue by theteladras** [basecamp/bc3-api#335]
Is there a way to verify if a webhook event trigger is coming from Basecamp?

There seems not to be any details mentioned on the [documentation](https://github.com/basecamp/bc3-api/blob/master/sections/webhooks.md#destroy-a-webhook) about it, and its kinda a requirement for any one who decides to integrate.

cc @jorgemanrubia

**Comment by jorgemanrubia** [comment]
@theteladras the `User-Agent` request header will be `Basecamp3 Webhook`.

**Comment by nimrodolev** [comment]
@jorgemanrubia a `User-Agent` header isn't really a trust mechanism - as I'm sure you're aware, anyone can issue an HTTP request with any User-Agent value they wish, e.g.:

```sh
curl --user-agent 'Basecamp3 Webhook' 'http://very-trusting.site'
```

Now, if we assume that -
1. The Payload URL is only meant to be invoked by Basecamp, when there's a change in the data, and that ↴
2. Once the Payload URL is invoked, the integrated system performs actions based on the data in the payload, and that ↴
3. Actions triggered by a malicious actor can do bad, bad things to a system, and that ↴
4. Bad things are - well - bad, and we should try to avoid them.

Then I think it's quite clear that an integration to Basecamp that uses webhooks can't be assumed secure if critical actions happening within it are dependent on a public HTTP endpoint that has no way of validating who's invoking it.

No need to reinvent the wheel, you're not the first company to implement webhooks and there are ideas out there -
[Github webhooks use a signature created using a shared secret](https://docs.github.com/en/developers/webhooks-and-events/webhooks/securing-your-webhooks)
[Google allows passing a secret value on each webhook creation](https://developers.google.com/drive/api/guides/push#:~:text=A%20token%20property,length%3A%0A%20%20%20%20%20%20%20%20%20%20256%20characters.)
And I've even remember enountring APIs that just give you a fixed set of IP ranges to verify the request is coming from the correct set of servers - though admittedly, I can't find an example of that at the moment.

So, to summarise, could we get something better than a UA header?

**Comment by jorgemanrubia** [comment]
Sorry @nimrodolev, you are right. A mechanism to sign webhooks requests is on our radar. It's something we just have to add 👍. Sorry I interpreted your question as looking for a discrimination mechanism, not an authentication one. My bad for reading quickly.
